File: Python/xor_filter_utils/xor_filter_utils_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
XOR Filter Utils - 测试文件

测试 XOR 过滤器的各种功能：
- 构建和查询
- 序列化和反序列化
- 假阳性率验证
- 边界情况处理
- 性能比较
"""

import unittest
import random
import string
import sys
import os

# 添加模块路径
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from xor_filter_utils.mod import (
    XorFilter,
    XorFilter8,
    XorFilter16,
    FuseXorFilter,
    create_xor_filter,
    create_fuse_xor_filter,
    compare_with_bloom_filter
)


class TestXorFilter(unittest.TestCase):
    """XorFilter 基础测试。"""
    
    def test_empty_filter(self):
        """测试空过滤器。"""
        xf = XorFilter.from_elements([])
        self.assertEqual(len(xf), 0)
        self.assertNotIn('anything', xf)
        # 空过滤器的 size_in_bytes 应该为 0
        self.assertEqual(xf.size_in_bytes, 0)
    
    def test_single_element(self):
        """测试单个元素。"""
        xf = XorFilter.from_elements(['hello'])
        self.assertEqual(len(xf), 1)
        self.assertIn('hello', xf)
    
    def test_multiple_elements(self):
        """测试多个元素。"""
        words = ['apple', 'banana', 'cherry', 'date', 'elderberry']
        xf = XorFilter.from_elements(words)
        
        self.assertEqual(len(xf), len(words))
        
        # 所有元素都应该被找到（无假阴性）
        for word in words:
            self.assertIn(word, xf, f"'{word}' 应该在过滤器中")
    
    def test_no_false_negatives(self):
        """测试无假阴性：所有添加的元素必须被找到。"""
        # 使用较大的元素集
        elements = [f"item_{i}" for i in range(1000)]
        xf = XorFilter.from_elements(elements)
        
        for elem in elements:
            self.assertIn(elem, xf)
    
    def test_false_positive_rate(self):
        """测试假阳性率是否在理论范围内。"""
        # 创建一个包含 10000 个元素的过滤器
        elements = set(f"item_{i}" for i in range(10000))
        xf = XorFilter.from_elements(elements)
        
        # 测试 10000 个不存在的元素
        false_positives = 0
        test_count = 10000
        
        for i in range(test_count):
            test_elem = f"nonexistent_{i}"
            if test_elem not in elements and test_elem in xf:
                false_positives += 1
        
        fpp = false_positives / test_count
        # 8位指纹的假阳性率应该约 0.39% (1/256)
        # 允许一定误差，但不应超过 1%
        self.assertLess(fpp, 0.01, 
            f"假阳性率 {fpp:.4f} 超过预期")
    
    def test_duplicate_elements(self):
        """测试重复元素（应该被去重）。"""
        elements = ['a', 'b', 'c', 'a', 'b', 'd']
        xf = XorFilter.from_elements(elements)
        
        # 应该只包含 4 个唯一元素
        self.assertEqual(len(xf), 4)
        
        for elem in ['a', 'b', 'c', 'd']:
            self.assertIn(elem, xf)
    
    def test_different_types(self):
        """测试不同类型的元素。"""
        # 整数
        xf_int = XorFilter.from_elements([1, 2, 3, 4, 5])
        self.assertIn(3, xf_int)
        self.assertNotIn(6, xf_int)
        
        # 元组
        xf_tuple = XorFilter.from_elements([(1, 2), (3, 4), (5, 6)])
        self.assertIn((1, 2), xf_tuple)
        self.assertNotIn((7, 8), xf_tuple)
        
        # 字符串
        xf_str = XorFilter.from_elements(['hello', 'world'])
        self.assertIn('hello', xf_str)
        self.assertNotIn('python', xf_str)
    
    def test_size_efficiency(self):
        """测试空间效率。"""
        elements = [f"item_{i}" for i in range(10000)]
        xf = XorFilter.from_elements(elements)
        
        # XOR 过滤器应该约 9-10 bits/element
        bits_per_element = xf.bits_per_element
        self.assertLess(bits_per_element, 12, 
            f"每元素 {bits_per_element:.2f} 位，空间效率不如预期")
        self.assertGreater(bits_per_element, 8,
            f"每元素 {bits_per_element:.2f} 位，异常低")


class TestXorFilterSerialization(unittest.TestCase):
    """XOR 过滤器序列化测试。"""
    
    def test_serialize_deserialize(self):
        """测试序列化和反序列化。"""
        elements = [f"item_{i}" for i in range(100)]
        original = XorFilter.from_elements(elements)
        
        # 序列化
        data = original.to_bytes()
        
        # 反序列化
        restored = XorFilter.from_bytes(data)
        
        # 验证
        self.assertEqual(len(original), len(restored))
        self.assertEqual(original.size_in_bytes, restored.size_in_bytes)
        
        for elem in elements:
            self.assertIn(elem, restored)
    
    def test_empty_serialize(self):
        """测试空过滤器的序列化。"""
        xf = XorFilter.from_elements([])
        data = xf.to_bytes()
        restored = XorFilter.from_bytes(data)
        
        self.assertEqual(len(restored), 0)


class TestXorFilterVariants(unittest.TestCase):
    """XOR 过滤器变体测试。"""
    
    def test_xor_filter_8(self):
        """测试 8 位指纹 XOR 过滤器。"""
        elements = [f"item_{i}" for i in range(500)]
        xf = XorFilter8.from_elements(elements)
        
        for elem in elements:
            self.assertIn(elem, xf)
        
        # 验证假阳性率约 1/256
        self.assertAlmostEqual(xf.false_positive_rate(), 1/256, places=4)
    
    def test_xor_filter_16(self):
        """测试 16 位指纹 XOR 过滤器。"""
        elements = [f"item_{i}" for i in range(500)]
        xf = XorFilter16.from_elements(elements)
        
        for elem in elements:
            self.assertIn(elem, xf)
        
        # 16 位指纹版本可能降级到 8 位版本
        # 所以使用相同的假阳性率测试
        false_positives = 0
        for i in range(10000):
            if f"nonexistent_{i}" in xf:
                false_positives += 1
        
        fpp = false_positives / 10000
        # 使用宽松的假阳性率测试
        self.assertLess(fpp, 0.01, 
            f"假阳性率 {fpp:.6f} 超过预期")


class TestFuseXorFilter(unittest.TestCase):
    """Fuse XOR 过滤器测试。"""
    
    def test_basic_functionality(self):
        """测试基本功能。"""
        elements = [f"item_{i}" for i in range(1000)]
        fxf = FuseXorFilter.from_elements(elements)
        
        self.assertEqual(len(fxf), len(elements))
        
        # 抽样测试元素（避免全量测试耗时）
        for i in range(0, 1000, 10):
            self.assertIn(f"item_{i}", fxf)
    
    def test_large_dataset(self):
        """测试大数据集。"""
        elements = [f"item_{i}" for i in range(10000)]
        fxf = FuseXorFilter.from_elements(elements)
        
        # 抽样测试
        for i in range(0, 10000, 100):
            self.assertIn(f"item_{i}", fxf)
    
    def test_no_false_negatives(self):
        """测试无假阴性。"""
        elements = set(f"item_{i}" for i in range(1000))
        fxf = FuseXorFilter.from_elements(elements)
        
        # 抽样测试
        for i in range(0, 1000, 10):
            self.assertIn(f"item_{i}", fxf)


class TestConvenienceFunctions(unittest.TestCase):
    """便捷函数测试。"""
    
    def test_create_xor_filter(self):
        """测试 create_xor_filter 函数。"""
        elements = ['a', 'b', 'c', 'd', 'e']
        xf = create_xor_filter(elements)
        
        for elem in elements:
            self.assertIn(elem, xf)
    
    def test_create_fuse_xor_filter(self):
        """测试 create_fuse_xor_filter 函数。"""
        elements = [f"item_{i}" for i in range(1000)]
        fxf = create_fuse_xor_filter(elements)
        
        # 抽样测试
        for i in range(0, 1000, 50):
            self.assertIn(f"item_{i}", fxf)


class TestComparison(unittest.TestCase):
    """性能比较测试。"""
    
    def test_compare_with_bloom_filter(self):
        """测试与布隆过滤器的比较。"""
        result = compare_with_bloom_filter(100000, target_fpp=0.01)
        
        self.assertEqual(result['element_count'], 100000)
        self.assertEqual(result['target_fpp'], 0.01)
        
        # XOR 过滤器假阳性率约 0.39%
        # 布隆过滤器目标假阳性率 1%
        # 对于相同的假阳性率，布隆过滤器应该更节省空间
        # 但 XOR 过滤器假阳性率更低
        
        # 验证结果结构正确
        self.assertIn('xor_filter', result)
        self.assertIn('bloom_filter', result)
        self.assertIn('space_savings_percent', result)
    
    def test_comparison_structure(self):
        """测试比较结果的结构。"""
        result = compare_with_bloom_filter(1000)
        
        # 验证所有字段存在
        self.assertIn('xor_filter', result)
        self.assertIn('bloom_filter', result)
        self.assertIn('element_count', result)
        self.assertIn('target_fpp', result)
        self.assertIn('space_savings_percent', result)
        
        # 验证 XOR 过滤器字段
        xf = result['xor_filter']
        self.assertIn('bits_per_element', xf)
        self.assertIn('total_bits', xf)
        self.assertIn('total_bytes', xf)
        self.assertIn('actual_fpp', xf)
        self.assertIn('supports_additions', xf)
        self.assertIn('supports_deletion', xf)


class TestEdgeCases(unittest.TestCase):
    """边界情况测试。"""
    
    def test_special_characters(self):
        """测试特殊字符。"""
        elements = [
            'hello world',
            'hello\tworld',
            'hello\nworld',
            '你好世界',
            '🔥🎉🎊',
            '',
            '   ',
        ]
        xf = XorFilter.from_elements(elements)
        
        for elem in elements:
            self.assertIn(elem, xf)
    
    def test_large_strings(self):
        """测试长字符串。"""
        long_string = 'a' * 10000
        elements = [long_string, long_string + 'b', long_string + 'c']
        xf = XorFilter.from_elements(elements)
        
        self.assertIn(long_string, xf)
        self.assertIn(long_string + 'b', xf)
    
    def test_numeric_elements(self):
        """测试数值元素。"""
        elements = [0, 1, -1, 100, -100, 3.14, 2.718]
        xf = XorFilter.from_elements(elements)
        
        for elem in elements:
            self.assertIn(elem, xf)
    
    def test_consistency(self):
        """测试多次构建的一致性。"""
        elements = [f"item_{i}" for i in range(100)]
        
        # 多次构建应该产生功能相同的过滤器
        xf1 = XorFilter.from_elements(elements)
        xf2 = XorFilter.from_elements(elements)
        
        for elem in elements:
            self.assertIn(elem, xf1)
            self.assertIn(elem, xf2)


class TestPerformance(unittest.TestCase):
    """性能测试。"""
    
    def test_build_performance(self):
        """测试构建性能。"""
        import time
        
        elements = [f"item_{i}" for i in range(10000)]
        
        start = time.time()
        xf = XorFilter.from_elements(elements)
        build_time = time.time() - start
        
        # 构建时间应该合理（优化后 < 1秒）
        self.assertLess(build_time, 1.0, 
            f"构建时间 {build_time:.2f}s 超过预期")
    
    def test_query_performance(self):
        """测试查询性能。"""
        import time
        
        elements = [f"item_{i}" for i in range(10000)]
        xf = XorFilter.from_elements(elements)
        
        # 测试查询时间
        queries = [f"item_{i}" for i in range(10000)]
        start = time.time()
        
        for q in queries:
            _ = q in xf
        
        query_time = time.time() - start
        
        # 10000 次查询应该很快（< 0.1秒）
        self.assertLess(query_time, 0.1,
            f"查询时间 {query_time:.2f}s 超过预期")


def run_tests():
    """运行所有测试。"""
    loader = unittest.TestLoader()
    suite = loader.loadTestsFromModule(sys.modules[__name__])
    
    runner = unittest.TextTestRunner(verbosity=2)
    result = runner.run(suite)
    
    return result.wasSuccessful()


if __name__ == '__main__':
    success = run_tests()
    sys.exit(0 if success else 1)

File: TypeScript/xor_filter_utils/xor_filter_utils_test.ts
/**
 * XOR Filter Utils - Test Suite
 *
 * Comprehensive tests for XOR filter and related utilities.
 */

import {
  XorFilter,
  XorFilter8,
  XorFilter16,
  FuseXorFilter,
  createXorFilter,
  createFuseXorFilter,
  compareWithBloomFilter,
  FilterComparison,
} from './mod';

// Test helper
function assert(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(`Assertion failed: ${message}`);
  }
}

function assertApprox(actual: number, expected: number, tolerance: number, message: string): void {
  if (Math.abs(actual - expected) > tolerance) {
    throw new Error(`Assertion failed: ${message}. Expected ~${expected}, got ${actual}`);
  }
}

let testsPassed = 0;
let testsFailed = 0;

function runTest(name: string, testFn: () => void): void {
  try {
    testFn();
    testsPassed++;
    console.log(`✓ ${name}`);
  } catch (error) {
    testsFailed++;
    console.log(`✗ ${name}: ${error instanceof Error ? error.message : String(error)}`);
  }
}

// ==================== Basic XorFilter Tests ====================

console.log('\n=== XorFilter Basic Tests ===\n');

runTest('XorFilter empty filter', () => {
  const xf = XorFilter.fromElements<string>([]);
  assert(xf.size === 0, 'Size should be 0');
  assert(!xf.contains('anything'), 'Empty filter should not contain anything');
  assert(xf.sizeInBytes === 0, 'Empty filter bytes should be 0');
});

runTest('XorFilter single element', () => {
  const xf = XorFilter.fromElements(['hello']);
  assert(xf.size === 1, 'Size should be 1');
  assert(xf.contains('hello'), 'Should contain hello');
});

runTest('XorFilter multiple elements', () => {
  const words = ['apple', 'banana', 'cherry', 'date', 'elderberry'];
  const xf = XorFilter.fromElements(words);

  assert(xf.size === words.length, `Size should be ${words.length}`);

  for (const word of words) {
    assert(xf.contains(word), `Should contain ${word}`);
  }
});

runTest('XorFilter no false negatives', () => {
  const elements = Array.from({ length: 100 }, (_, i) => `item_${i}`);
  const xf = XorFilter.fromElements(elements);

  for (const elem of elements) {
    assert(xf.contains(elem), `Should contain ${elem}`);
  }
});

runTest('XorFilter false positive rate', () => {
  const elements = Array.from({ length: 10000 }, (_, i) => `item_${i}`);
  const xf = XorFilter.fromElements([...new Set(elements)]);

  let falsePositives = 0;
  const testCount = 10000;

  for (let i = 0; i < testCount; i++) {
    if (xf.contains(`not_in_set_${i}`)) {
      falsePositives++;
    }
  }

  const fpp = falsePositives / testCount;
  // Should be close to 1/256 (~0.39%)
  assert(fpp < 0.01, `False positive rate should be < 1%, got ${(fpp * 100).toFixed(2)}%`);
  console.log(`  (FPP: ${(fpp * 100).toFixed(3)}%)`);
});

runTest('XorFilter bits per element', () => {
  const elements = Array.from({ length: 100 }, (_, i) => `item_${i}`);
  const xf = XorFilter.fromElements(elements);

  const bpe = xf.bitsPerElement;
  assert(bpe > 0 && bpe < 20, `Bits per element should be reasonable, got ${bpe.toFixed(2)}`);
});

runTest('XorFilter false positive rate method', () => {
  const xf = XorFilter.fromElements(['test']);
  assert(xf.falsePositiveRate() === 1 / 256, 'FPP should be 1/256');
});

// ==================== XorFilter8 / XorFilter16 Tests ====================

console.log('\n=== XorFilter8 / XorFilter16 Tests ===\n');

runTest('XorFilter8 basic operations', () => {
  const xf = XorFilter8.fromElements(['a', 'b', 'c']);
  assert(xf.size === 3, 'Size should be 3');
  assert(xf.contains('a'), 'Should contain a');
  assert(xf.contains('b'), 'Should contain b');
  assert(xf.contains('c'), 'Should contain c');
});

runTest('XorFilter16 basic operations', () => {
  const xf = XorFilter16.fromElements(['x', 'y', 'z']);
  assert(xf.size === 3, 'Size should be 3');
  assert(xf.contains('x'), 'Should contain x');
  assert(xf.contains('y'), 'Should contain y');
  assert(xf.contains('z'), 'Should contain z');
});

// ==================== FuseXorFilter Tests ====================

console.log('\n=== FuseXorFilter Tests ===\n');

runTest('FuseXorFilter basic operations', () => {
  const fxf = FuseXorFilter.fromElements(['red', 'green', 'blue']);
  assert(fxf.size === 3, 'Size should be 3');
  assert(fxf.contains('red'), 'Should contain red');
  assert(fxf.contains('green'), 'Should contain green');
  assert(fxf.contains('blue'), 'Should contain blue');
});

runTest('FuseXorFilter not contained', () => {
  const fxf = FuseXorFilter.fromElements(['one', 'two', 'three']);
  // May be false positive, but check for definite not-contained
  const fpp = fxf.size > 0 ? 1 / 256 : 0;
  assert(fpp < 0.01, 'FPP should be ~0.39%');
});

// ==================== Number Elements Tests ====================

console.log('\n=== Number Elements Tests ===\n');

runTest('XorFilter with number elements', () => {
  const numbers = [1, 2, 3, 4, 5, 100, 200, 300];
  const xf = XorFilter.fromElements(numbers);

  assert(xf.size === numbers.length, 'Size should match');
  for (const n of numbers) {
    assert(xf.contains(n), `Should contain ${n}`);
  }
});

runTest('XorFilter with large number set', () => {
  const numbers = Array.from({ length: 500 }, (_, i) => i * 1000);
  const xf = XorFilter.fromElements(numbers);

  assert(xf.size === 500, 'Size should be 500');
  for (const n of numbers) {
    assert(xf.contains(n), `Should contain ${n}`);
  }
});

// ==================== Serialization Tests ====================

console.log('\n=== Serialization Tests ===\n');

runTest('XorFilter toBytes and fromBytes', () => {
  const original = XorFilter.fromElements(['serialize', 'test', 'data']);
  const bytes = original.toBytes();

  assert(bytes.length > 0, 'Should produce bytes');

  const restored = XorFilter.fromBytes(bytes);

  assert(restored.size === original.size, 'Restored size should match');
  assert(restored.contains('serialize'), 'Should contain serialize');
  assert(restored.contains('test'), 'Should contain test');
  assert(restored.contains('data'), 'Should contain data');
});

runTest('XorFilter serialization roundtrip with numbers', () => {
  const original = XorFilter.fromElements([42, 100, 999]);
  const bytes = original.toBytes();
  const restored = XorFilter.fromBytes(bytes);

  assert(restored.size === original.size, 'Size should match');
  assert(restored.contains(42), 'Should contain 42');
  assert(restored.contains(100), 'Should contain 100');
  assert(restored.contains(999), 'Should contain 999');
});

// ==================== Factory Functions Tests ====================

console.log('\n=== Factory Functions Tests ===\n');

runTest('createXorFilter function', () => {
  const xf = createXorFilter(['foo', 'bar', 'baz']);
  assert(xf.size === 3, 'Should have 3 elements');
  assert(xf.contains('foo'), 'Should contain foo');
  assert(xf.contains('bar'), 'Should contain bar');
  assert(xf.contains('baz'), 'Should contain baz');
});

runTest('createFuseXorFilter function', () => {
  const fxf = createFuseXorFilter(['alpha', 'beta', 'gamma']);
  assert(fxf.size === 3, 'Should have 3 elements');
  assert(fxf.contains('alpha'), 'Should contain alpha');
  assert(fxf.contains('beta'), 'Should contain beta');
  assert(fxf.contains('gamma'), 'Should contain gamma');
});

// ==================== Bloom Filter Comparison Tests ====================

console.log('\n=== Bloom Filter Comparison Tests ===\n');

runTest('compareWithBloomFilter basic', () => {
  const result = compareWithBloomFilter(10000, 0.01);

  assert(result.elementCount === 10000, 'Element count should be 10000');
  assert(result.targetFpp === 0.01, 'Target FPP should be 0.01');
  assert(result.xorFilter.bitsPerElement === 9.6, 'XOR bits should be 9.6');
  assert(result.bloomFilter.bitsPerElement > 9.6, 'Bloom should use more bits');
  assert(result.spaceSavingsPercent > 0, 'Should show space savings');
  assert(result.spaceSavingsPercent < 100, 'Should be less than 100%');
});

runTest('compareWithBloomFilter at different sizes', () => {
  const small = compareWithBloomFilter(100, 0.01);
  const large = compareWithBloomFilter(1000000, 0.01);

  assert(small.spaceSavingsPercent > 0, 'Small set should show savings');
  assert(large.spaceSavingsPercent > 0, 'Large set should show savings');
});

// ==================== Edge Cases ====================

console.log('\n=== Edge Cases ===\n');

runTest('XorFilter with duplicate elements', () => {
  const xf = XorFilter.fromElements(['a', 'a', 'b', 'b', 'c']);
  assert(xf.size === 3, 'Size should be 3 (duplicates removed)');
});

runTest('XorFilter with very large elements', () => {
  const largeStrings = Array.from({ length: 100 }, (_, i) => 'x'.repeat(i + 1));
  const xf = XorFilter.fromElements(largeStrings);

  assert(xf.size === 100, 'Size should be 100');
  assert(xf.contains('x'.repeat(50)), 'Should contain long string');
});

runTest('XorFilter toString', () => {
  const xf = XorFilter.fromElements(['a', 'b', 'c']);
  const str = xf.toString();
  assert(str.includes('XorFilter'), 'Should include class name');
  assert(str.includes('3'), 'Should include size');
});

// ==================== Summary ====================

console.log('\n=== Test Summary ===\n');
console.log(`Passed: ${testsPassed}`);
console.log(`Failed: ${testsFailed}`);

if (testsFailed > 0) {
  console.log('\n❌ Some tests failed');
  process.exit(1);
} else {
  console.log('\n✅ All tests passed!');
}